core/src/test/scala/org/apache/spark/deploy/worker/WorkerSuite.scala

```diff
@@ -243,4 +244,59 @@ class WorkerSuite extends SparkFunSuite with Matchers with BeforeAndAfter {
       ExecutorStateChanged("app1", 0, ExecutorState.EXITED, None, None))
     assert(cleanupCalled.get() == value)
   }
+
+  test("removes our metadata of all executors registered for the given " +
+    "application after WorkDirCleanup when config spark.shuffle.service.db.enabled=true") {
+    testWorkDirCleanupAndRemoveMetadataWithConfig(true)
+  }
+
+  test("don't removes our metadata of all executors registered for the given " +
+    "application after WorkDirCleanup when config spark.shuffle.service.db.enabled=false") {
+    testWorkDirCleanupAndRemoveMetadataWithConfig(false)
+  }
+
+  private def testWorkDirCleanupAndRemoveMetadataWithConfig(value: Boolean) = {
+    val conf = new SparkConf().set("spark.shuffle.service.db.enabled", value.toString)
+    conf.set("spark.worker.cleanup.appDataTtl", 60.toString)
+    conf.set("spark.shuffle.service.enabled", "true")
+
+    val appId = "app1"
+    val execId = "exec1"
+    val cleanupCalled = new AtomicBoolean(false)
+    when(shuffleService.applicationRemoved(any[String])).thenAnswer(new Answer[Unit] {
+      override def answer(invocations: InvocationOnMock): Unit = {
+        cleanupCalled.set(true)
+      }
+    })
+    val externalShuffleServiceSupplier = new Supplier[ExternalShuffleService] {
+      override def get: ExternalShuffleService = shuffleService
+    }
+    val worker = makeWorker(conf, externalShuffleServiceSupplier)
+
+    val workDir = Option("/tmp/work").map(new File(_)).get
+    try {
+      workDir.mkdirs()
+      if (!workDir.exists() || !workDir.isDirectory) {
+        logError("Failed to create work directory " + workDir)
+      }
+      assert(workDir.isDirectory)
+    } catch {
+      case e: Exception =>
+        logError("Failed to create work directory " + workDir, e)
+    }
+    // initialize workers
+    worker.workDir = workDir
+    // Create the executor's working directory
+    val executorDir = new File(worker.workDir, appId + "/" + execId)
+
+    if (!executorDir.exists()) {
+      if (!executorDir.mkdirs()) {
+        throw new IOException("Failed to create directory " + executorDir)
+      }
+    }
+    executorDir.setLastModified(System.currentTimeMillis - (1000 * 120))
+    worker.receive(WorkDirCleanup)
+    Thread.sleep(10)
```

Comment:
Is there a way to avoid this sleep? On my machine running the test via IntelliJ this 10 milliseconds was not enough.

Reply:
Maybe overkill but moving in the Worker extracting the cleanup functionality (the cleanup Future body) into a separate method visible for the test and calling that function directly instead of sending WorkDirCleanup would solve this issue. Otherwise I would a bit worried about having a flaky test depending on Jenkins workload. What is your opinion?

Reply:
I just read @squito's solution for this problem and I like that as it is less intrusive.